### PR TITLE
[mqtt] Make external transport to await until broker is ready to serve clients

### PR DESCRIPTION
**Pull request by dmolokanov**
Make external transports to await special signal before starting to serve external clients.

### CODE REVIEW

File: mqtt/mqtt-bridge/Cargo.toml [hunk 1]
```diff
@@ -15,4 +15,6 @@ serial_test = "0.4"
 tracing = "0.1"
 
 edgelet-client = { path = "../edgelet-client" }
+
+[dev-dependencies]
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-bridge/Cargo.toml:19]: :+1:

File: mqtt/mqttd/src/broker/bootstrap/edgehub.rs [hunk 2]
```diff
@@ -92,12 +94,14 @@ where
 
     // Add system transport to allow communication between edgehub components
     let authenticator = LocalAuthenticator::new();
-    server.tcp(config.listener().system().addr(), authenticator);
+    server.tcp(config.listener().system().addr(), authenticator, None)?;
 
     // Add regular MQTT over TCP transport
     let authenticator = EdgeHubAuthenticator::new(config.auth().url());
+    let (tx, _) = broadcast::channel(1);
```
Comment (vadim-kovalyov) [inline mqtt/mqttd/src/broker/bootstrap/edgehub.rs:101]: nit:
```suggestion
    let (broker_ready, _) = broadcast::channel(1);
```

**Review by vadim-kovalyov — APPROVED** [review]

File: mqtt/mqtt-broker/src/server.rs [hunk 3]
```diff
@@ -225,71 +232,124 @@ where
...
-                        }
-                    });
-                }
-                Either::Left(_) => {
-                    info!(
-                        "accept loop shutdown. no longer accepting connections on {}",
-                        addr
-                    );
-                    break;
+struct Listener {
```
Comment (davilu) [inline mqtt/mqtt-broker/src/server.rs:235]: nit: rename to `ConnectionListener`

**Review by davilu — APPROVED** [review]

File: mqtt/mqttd/src/broker/bootstrap/edgehub.rs [hunk 4]
```diff
@@ -131,6 +138,12 @@ where
     pin_mut!(renewal_signal);
     let shutdown = future::select(shutdown_signal, renewal_signal).map(drop);
 
+    // TODO remove this call when broker readiness is implemented
+    tokio::spawn(async move {
```
Comment (davilu) [inline mqtt/mqttd/src/broker/bootstrap/edgehub.rs:142]: Is `broker_ready`  going to be passed into some `CommandHandler` to trigger the signal?

File: mqtt/mqtt-broker/src/server.rs [hunk 5]
```diff
@@ -225,71 +232,124 @@ where
...
                 }
-                Either::Right((Some(Err(e)), _)) => {
-                    warn!(
-                        message = "accept loop exiting due to an error",
-                        error =% DetailedErrorValue(&e)
-                    );
-                    break;
+                None => Either::Right(future::ready(Ok(()))),
+            };
+            pin_mut!(ready);
```
Comment (varunpuranik) [inline mqtt/mqtt-broker/src/server.rs:292]: Why is this needed?

Comment (dmolokanov) [reply mqtt/mqtt-broker/src/server.rs:292]: `future::select` one line below expects some type F that implements both `Future` and `Unpin`. `async` block does not impl `Unpin` so we have to make them pinned

Comment (dmolokanov) [reply mqtt/mqtt-broker/src/server.rs:292]: and we need `async` block to consume `ready` receiver. borrow checker complains otherwise

File: mqtt/mqtt-broker/src/server.rs [hunk 6]
```diff
@@ -51,36 +47,53 @@ where
     Z: Authorizer + Send + 'static,
     P: MakeIncomingPacketProcessor + MakeOutgoingPacketProcessor + Clone + Send + Sync + 'static,
 {
-    pub fn tcp<N>(&mut self, addr: &str, authenticator: N) -> &mut Self
+    pub fn tcp<A, N>(
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-broker/src/server.rs:50]: nit: looks like a builder pattern? maybe rename to `with_tcp`, etc..?

File: mqtt/mqtt-broker/src/server.rs [hunk 7]
```diff
@@ -225,71 +232,124 @@ where
...
+            {
+                Either::Left((Ok(_), mut shutdown_signal)) => {
+                    // start listening incoming connections on given network address
+                    let mut incoming = transport.incoming().await?;
+
+                    let addr = incoming.local_addr()?;
+                    info!("Listening on address {}", addr);
+
+                    loop {
+                        match future::select(&mut shutdown_signal, incoming.next()).await {
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-broker/src/server.rs:305]: oh, this is interesting `&mut shutdown_signal`.... I always had problems with `future::select` in a loop, b/c it consumed futures in every iteration. Did not know that you can pass mut ref.

**Review by vadim-kovalyov — APPROVED** [review]
:+1:

File: mqtt/mqtt-broker/src/server.rs (hunk 5 above)
Comment (varunpuranik) [reply mqtt/mqtt-broker/src/server.rs:292]: got it, thx.
